wenet/transformer/encoder.py

```diff
@@ -163,13 +219,43 @@ def forward(
             xs = self.global_cmvn(xs)
         xs, pos_emb, masks = self.embed(xs, masks)
         mask_pad = masks  # (B, 1, T/subsample_rate)
-        chunk_masks = add_optional_chunk_mask(xs, masks,
-                                              self.use_dynamic_chunk,
-                                              self.use_dynamic_left_chunk,
-                                              decoding_chunk_size,
-                                              self.static_chunk_size,
-                                              num_decoding_left_chunks)
-        for layer in self.encoders:
+        chunk_masks = add_optional_chunk_mask(
+            xs,
+            masks,
+            self.use_dynamic_chunk,
+            self.use_dynamic_left_chunk,
+            decoding_chunk_size,
+            self.static_chunk_size,
+            num_decoding_left_chunks,
+        )
+
+        time_reduce_residuals = torch.jit.annotate(
+            List[Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]], []
+        )
+        time_reduce_level: int = 0
+        for i, layer in enumerate(self.encoders):
+            if len(self.time_reduce_idx) > 0 and i in self.time_reduce_idx:
+                time_reduce_residuals.append((xs, mask_pad, chunk_masks, pos_emb))
+                for i, reduce_layer in enumerate(self.time_reduce_layers):
+                    if time_reduce_level == i:
+                        xs, mask_pad, chunk_masks, pos_emb = reduce_layer(
+                            xs, mask_pad, chunk_masks, pos_emb
+                        )
+                time_reduce_level += 1
+            if len(self.time_reduce_idx) > 0 and i in self.time_recover_idx:
+                time_reduce_level -= 1
+                residual_xs, mask_pad, chunk_masks, pos_emb = time_reduce_residuals[
+                    time_reduce_level
+                ]
+                B, T_, D = xs.size()
+                xs = xs.unsqueeze(2).tile((1, 1, 2, 1))  # (B,T',2,D')
+                xs = xs.reshape(B, -1, D)[
+                    :, : residual_xs.size(1), :
+                ]  # (B,T,D'), T=floor(T*2)
+                for i, recover_layer in enumerate(self.time_recover_layers):
+                    if time_reduce_level == i:
+                        xs = recover_layer(xs)
+                xs += residual_xs  # (B,T,D)
             xs, chunk_masks, _, _ = layer(xs, chunk_masks, pos_emb, mask_pad)
         if self.normalize_before:
             xs = self.after_norm(xs)
```

Comment:
It should be  a `PreLN` at the beginning of blocks if you use `Squeezeformer`. This `after_norm` would be duplicate because there are `PostLN` at the end of every `Squeezeformer` block.

Reply:
As `Squeezeformer` always employ `PostLN` rather than `PreLN`, there's no option to choose `PreLN`.
Therefore, to avoid such a duplication problem, I just inserted an `assert` code to make sure `normalize_before` is `False` in `SqueezeformerEncoder`.